mbedtls/build.rs

```diff
@@ -6,32 +6,49 @@
  * option. This file may not be copied, modified, or distributed except
  * according to those terms. */
 
+use std::collections::hash_map::DefaultHasher;
 use std::collections::{HashMap, HashSet};
-use std::env;
+use std::hash::Hasher;
 
 use rustc_version::Channel;
+use std::env;
 
-/// Return the crate hash that Cargo will be passing to `rustc -C metadata=`.
-// If there's a panic in this code block, that means Cargo's way of running the
-// build script has changed, and this code should be updated to handle the new
-// case.
-fn get_compilation_metadata_hash() -> String {
+/// Retrieves or generates a metadata value used for symbol name mangling to ensure unique C symbols.
+/// When building with Cargo, the metadata value is extracted from the OUT_DIR environment variable.
+/// For Bazel builds, this method approximates Cargo's -Cmetadata by hashing the crate name and version,
+/// which are sufficient for ensuring symbol uniqueness.
+fn get_compilation_symbol_suffix() -> String {
     let out_dir: std::path::PathBuf = std::env::var_os("OUT_DIR").unwrap().into();
     let mut out_dir_it = out_dir.iter().rev();
-    assert_eq!(out_dir_it.next().unwrap(), "out");
-    let crate_ = out_dir_it.next().unwrap().to_string_lossy();
-    assert!(crate_.starts_with("mbedtls-"));
-    crate_[8..].to_owned()
+    let next = out_dir_it.next().unwrap();
+
+    if next == "out" {
+        // If Cargo is used as build system.
+        let crate_ = out_dir_it.next().unwrap().to_string_lossy();
+        assert!(crate_.starts_with("mbedtls-"));
+        return crate_[8..].to_owned();
+    } else if next == "_bs.out_dir" {
```

Comment:
Is this check correct?

When I do a build using Bazel, my OUT_DIR env var is set as, e.g.,:
```
OUT_DIR='${pwd}/bazel-out/k8-fastbuild/bin/external/crate_index__mbedtls-sys-auto-2.26.1/mbedtls-sys-auto_bs.out_dir'
```
There is no `_bs.out_dir` component in that path.

Reply:
I've never seen that format. In my setup with the latest version of bazel/Rules_rust the OUT_DIR is in this format:
`OUT_DIR='${pwd}/bazel-out/k8-fastbuild/bin/external/crate_index__mbedtls-0.12.3/_bs.out_dir' \`
Could you please provide your WORKSPACE/Branch to reproduce it?